### ISSUE TITLE
launch.json can't find prelaunch task

### ISSUE THREAD

**Issue by vescovd1** [ARM-software/vscode-cmsis-csolution#57]
Type: Bug Report

**Describe the bug**
- OS and Version: Windows 10
- VS Code Version: 1.87.2
- Extension Version:  v1.83.0
- Target Device: STM32
- Other extensions you installed (and if the issue persists after disabling them):
- A clear and concise description of what the bug is.  The launch.json cannot find the predefined prelaunch task. This use to work but is no longer functional on my setup. 

**To Reproduce**

**Expected behavior**

**Code sample and logs**
- Code sample
- `launch.json: 
- {
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Arm Debug",
            "type": "arm-debug",
            "request": "launch",
            "serialNumber": "${command:device-manager.getSerialNumber}",
            "program": "${command:embedded-debug.getApplicationFile}",
            "cmsisPack": "${command:cmsis-csolution.getTargetPack}",
            "connectMode": "auto",
            "debugFrom": "main",
            "probe": "CMSIS-DAP",
            "programMode": "auto",
            "resetMode": "auto",
            "preLaunchTask": "Flash Device"
        }
    ]
}
task.json:
{
	"version": "2.0.0",
	"tasks": [
		{
			"label": "Build Context",
			"type": "cmsis-csolution.build",
			"solution": "${command:cmsis-csolution.getSolutionPath}",
			"rebuild": false,
			"problemMatcher": [],
			"group": {
				"kind": "build",
				"isDefault": true
			}
		},
		{
			"label": "Flash Device",
			"type": "arm-debugger.flash",
			"serialNumber": "${command:device-manager.getSerialNumber}",
			"program": "${command:arm-debugger.getApplicationFile}",
			"cmsisPack": "${command:device-manager.getDevicePack}",
			"problemMatcher": [],
			"dependsOn": "Build Context"
		}
	]
}

`

**Screenshots**

![image](https://github.com/ARM-software/vscode-cmsis-csolution/assets/111083048/32fa0b93-d385-4418-9577-b70dfc3f34f6)

**Additional context**
 I can trigger the individual Flash Device task and that works. Just not from launch.json

**Comment by vescovd1** [comment]
I am triggering from the "Debug" icon from the CMSIS view

**Comment by arekzaluski** [comment]
Hi @vescovd1,

Issue exists because your launch config uses [Arm Embedded Debugger extension](https://marketplace.visualstudio.com/items?itemName=Arm.embedded-debug) for debugging  and [Arm Debugger extension](https://marketplace.visualstudio.com/items?itemName=Arm.arm-debugger) for flashing the application on the device.
I recommend adding the new Debug configuration that uses Arm Debugger. You can find more on how to set it up here:
[Arm Debugger configuration](https://developer.arm.com/documentation/108029/0000/Arm-Debugger-extension/Debug-your-project-with-Arm-Debugger/Add-configuration?lang=en)

**Comment by vescovd1** [comment]
so I tried changing my launch json to arm-debugger from arm-debug but it still does not work.

launch.json:
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        
        {
            "name": "Arm Debugger",
            "type": "arm-debugger",
            "request": "launch",
            "serialNumber": "${command:device-manager.getSerialNumber}",
            "program": "${command:arm-debugger.getApplicationFile}",
            "cmsisPack": "${command:cmsis-csolution.getTargetPack}",
            "deviceName": "${command:cmsis-csolution.getDeviceName}",
            "processorName": "${command:cmsis-csolution.getProcessorName}",
            "preLaunchTask": "arm-debugger.flash: Flash Device"
        }
    ]
}

tasks.json:
{
	"version": "2.0.0",
	"tasks": [
		{
			"label": "arm-debugger.flash: Flash Device",
			"type": "arm-debugger.flash",
			"serialNumber": "${command:device-manager.getSerialNumber}",
			"program": "${command:arm-debugger.getApplicationFile}",
			"cmsisPack": "${command:cmsis-csolution.getTargetPack}",
			"deviceName": "${command:cmsis-csolution.getDeviceName}",
			"processorName": "${command:cmsis-csolution.getProcessorName}",
			"problemMatcher": []
		}
	]
}
![cantfind](https://github.com/ARM-software/vscode-cmsis-csolution/assets/111083048/9ed1b382-83fb-477f-a5e2-761d0587a68f)

**Comment by vescovd1** [comment]
all I really want to do is press my debug icon in the cmsis context window and have it launch the flash, and flash should have a depends on "build" task.  Please tell how to do this as the whole ARM: debug, ARM: debugger and ARM: embedded debugger is very confusing and poorly documented.

**Comment by vescovd1** [comment]
with this configuration for task.json:
{
	"version": "2.0.0",
	"tasks": [
		{
			"label": "Build Context",
			"type": "cmsis-csolution.build",
			"solution": "${command:cmsis-csolution.getSolutionPath}",
			"rebuild": false,
			"problemMatcher": [],
			"group": {
				"kind": "build",
				"isDefault": true
			}
		},
		{
			"label": "arm-debugger.flash: Flash Device",
			"type": "arm-debugger.flash",
			"serialNumber": "${command:device-manager.getSerialNumber}",
			"program": "${command:arm-debugger.getApplicationFile}",
			"cmsisPack": "${command:cmsis-csolution.getTargetPack}",
			"deviceName": "${command:cmsis-csolution.getDeviceName}",
			"processorName": "${command:cmsis-csolution.getProcessorName}",
			"problemMatcher": [],
			"dependsOn": "Build Context"
		}
	]
}

I can MANUALLY trigger flash task and it works as expected. If build is out of date the build is pretriggered as expected.

I can also launch debugger after MANUALLY triggering flash load if I remove the "preLaunchTask" line from the launch json. Debugger attaches and works just fine. Its only when I add the "preLaunchTask" and point it to flash task does it not work.

**Comment by thegecko** [comment]
You may be hitting this issue:
https://github.com/microsoft/vscode/issues/157041

Can you try running the debug through the debug UI instead of the control in the outline view?

**Comment by vescovd1** [comment]
Don't think so as "dependsOn" works just fine. It's the "preLaunchTask" that is the problem.

**Comment by vescovd1** [comment]
so its only when you launch the debugger via the CMSIS view.  Doing the Run->Debug from top menu works and also by pressing F5 works.  See screen shot.
![debug launch](https://github.com/ARM-software/vscode-cmsis-csolution/assets/111083048/de7ecd31-1cdf-4b86-8ba6-70f4dd1bcf99)

**Comment by thegecko** [comment]
> so its only when you launch the debugger via the CMSIS view. Doing the Run->Debug from top menu works and also by pressing F5 works. See screen shot.

Yes, so it this bug: https://github.com/microsoft/vscode/issues/157041

I'll update it to mention it occurs with debug too.

**Comment by jkrech** [comment]
Marked duplicate, now tracked here: https://github.com/Open-CMSIS-Pack/vscode-cmsis-solution/issues/76
